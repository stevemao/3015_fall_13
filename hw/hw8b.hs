last' [] = error "last': no last' in []."
last' (x:xs) = if null xs then x else last' xs

ismem  x [] = False 
ismem x (y:ys) = x == y || ismem x ys

allaremem [] _ = True
allaremem (x:xs) ys = ismem x ys && allaremem xs ys

find x [] = error "find: not found."
find x ((y,z): yzs) = if x == y then z else find x yzs

