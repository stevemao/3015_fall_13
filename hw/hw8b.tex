\documentclass[11pt]{article}

\include{header}

\begin{document}
\homework{8}{21 September}

Recall the function {\tt{last}} which was defined by recursion on the structure of the list.

\begin{program*}
\> last [] = error "last: nil." \\
\> last (x:xs) = if null xs then x else last xs\\
\end{program*}

You need to implement the following functions by recursion on the structure of
at least one of the list arguments:

\begin{problem}
{\tt{ismem :: (Eq a) => a -> [a] -> Bool}} 
\begin{quotation}A call to the function {\tt{ismem}} of
the form {\tt{(ismem x ys)}} returns {\tt{True}} if {\tt{x}} is in the list
{\tt{ys}} and returns {\tt{False}} otherwise.
\end{quotation}
\end{problem}

\begin{problem}
{\tt{allaremem :: (Eq a) => [a] -> [a] -> Bool}}
\begin{quotation}
  A call to {\tt{allaremem}} of the form {\tt{(allaremem xs ys)}} returns
  {\tt{True}} if all the elements in the list {\tt{xs}} occur in the list
  {\tt{ys}} and returns {\tt{False}} otherwise. In analogy with subsets, you
  might think of this as a sublist function, it returns true if the first list
  is a sublist of the second (disregarding order.) You may find your
  implementation of {\tt{ismem}} useful in defining {\tt{allaremem}}.
\end{quotation}
\end{problem}

\begin{problem}
{\tt{find :: (Eq a) => a -> [(a, b)] -> b}}
\begin{quotation}
 A call of the form {\tt{(find x xys)}} returns {\tt{y}} (where {\tt{y}} has
  type {\tt{b}}) if there is a pair of the form {\tt{(x,y)}} in the list {\tt{xys}}
  and raises an exception otherwise.
Here's a big hint on {\tt{find}}, the left side of the definition of
find could appear as follows:

\begin{program*}
\> find x [] =  ... \\
\> find x ((y,z): yzs) = ...
\end{program*}
 Setting up the left side of the definition like this allows for pattern
matching on the structure of the pairs in the list argument.  
\end{quotation}
\end{problem}




\end{document}
% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "pdflatex hw8b"
% fill-column:79
% End:




