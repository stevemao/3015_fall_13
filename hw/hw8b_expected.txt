
Prelude> :l hw8
[1 of 1] Compiling Main             ( hw8.hs, interpreted )
Ok, modules loaded: Main.
*Main> :t ismem
ismem :: (Eq a) => a -> [a] -> Bool
*Main> :t allaremem
allaremem :: (Eq t) => [t] -> [t] -> Bool
*Main> :t find
find :: (Eq a) => a -> [(a, t)] -> t
*Main> ismem 1 []
False
*Main> ismem 1 [2]
False
*Main> ismem 1 [[1]]

<interactive>:1:6:
    No instance for (Num [t])
      arising from the literal `1' at <interactive>:1:6
    Possible fix: add an instance declaration for (Num [t])
    In the first argument of `ismem', namely `1'
    In the expression: ismem 1 [[1]]
    In the definition of `it': it = ismem 1 [[1]]
*Main> ismem [1] [[1]]
True
*Main> ismem 1 [1]
True
*Main> ismem 1 [3,2,1]
True
*Main> ismem 2 [3,2,1]
True
*Main> ismem 3 [3,2,1]
True
*Main> allaremem [1,2,3] [3,2,1]
True
*Main> allaremem [1,2] [3,2,1]
True
*Main> allaremem [1] [3,2,1]
True
*Main> allaremem [] [3,2,1]
True
*Main> allaremem [] []
True
*Main> allaremem [1] []
False
*Main> allaremem [1,2] [1,3]
False
*Main> allaremem [1,2] [3,1]
False
*Main> find 1 []
*** Exception: find: not found.
*Main> find 1 [('a',1),('b',2)]

<interactive>:1:5:
    No instance for (Num Char)
      arising from the literal `1' at <interactive>:1:5
    Possible fix: add an instance declaration for (Num Char)
    In the first argument of `find', namely `1'
    In the expression: find 1 [('a', 1), ('b', 2)]
    In the definition of `it': it = find 1 [('a', 1), ('b', 2)]
*Main> find 1 [(1,'a'),(2,'b')]
'a'
*Main> find 2 [(1,'a'),(2,'b')]
'b'
*Main> find 3 [(1,'a'),(2,'b')]
*** Exception: find: not found.
*Main> find 3 [(1,'a'),(3,'z'),(2,'b')]
'z'
*Main> find [1,2] [([1],"a"),(2,"b")]

<interactive>:1:23:
    No instance for (Num [t])
      arising from the literal `2' at <interactive>:1:23
    Possible fix: add an instance declaration for (Num [t])
    In the expression: 2
    In the expression: (2, "b")
    In the second argument of `find', namely `[([1], "a"), (2, "b")]'
*Main> find [1,2] [([1],"a"),([2],"b")]
"*** Exception: find: not found.
*Main> find [1,2] [([1],"a"),([2],"b"),([1,2],"ab")]
"ab"
*Main> find [1,2] [([1],"one"),([2],"two"),([1,2],"one two"),([3],"three")]
"one two"
*Main> find [1] [([1],"one"),([2],"two"),([1,2],"one two"),([3],"three")]
"one"
*Main> find [3] [([1],"one"),([2],"two"),([1,2],"one two"),([3],"three")]
"three"
*Main> 
